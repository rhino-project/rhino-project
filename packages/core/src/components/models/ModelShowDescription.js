import { useModelShowContext } from 'rhino/hooks/controllers';
import { useGlobalComponent } from 'rhino/hooks/overrides';
import ModelSection from './ModelSection';

export const ModelShowDescriptionBase = (props) => {
  const { renderPaths } = useModelShowContext();

  return (
    <ModelSection baseClassName="show-description">{renderPaths}</ModelSection>
  );
};

ModelShowDescriptionBase.propTypes = {};

const ModelShowDescription = (props) =>
  useGlobalComponent(ModelShowDescriptionBase, props);

export default ModelShowDescription;
