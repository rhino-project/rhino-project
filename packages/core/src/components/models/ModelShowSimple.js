import PropTypes from 'prop-types';
import { useModelShowController } from '../../hooks/controllers';
import ModelShowProvider from './ModelShowProvider';
import { useMemo } from 'react';
import { Spinner } from 'reactstrap';

export const ModelShowSimple = ({ children, fallback = true, ...props }) => {
  const controller = useModelShowController(props);
  const { isInitialLoading } = controller;

  // Fallback mirrors React 18 Suspense
  const renderFallback = useMemo(() => {
    if (!isInitialLoading || !fallback) return children;

    if (fallback === true) return <Spinner />;

    return fallback;
  }, [children, fallback, isInitialLoading]);

  return (
    <ModelShowProvider {...controller}>{renderFallback}</ModelShowProvider>
  );
};

ModelShowSimple.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ])
};

export default ModelShowSimple;
