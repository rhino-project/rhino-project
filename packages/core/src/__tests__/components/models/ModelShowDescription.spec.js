import ModelShowDescription from 'rhino/components/models/ModelShowDescription';
import ModelShowSimple from 'rhino/components/models/ModelShowSimple';
import { sharedModelTests } from './sharedModelTests';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render } from '@testing-library/react';
import rhinoConfig from 'rhino.config';

vi.spyOn(rhinoConfig, 'components', 'get').mockReturnValue({
  version: 1,
  components: {}
});

describe('ModelShowDescription', () => {
  const wrapper = ({ children }) => {
    const queryClient = new QueryClient({
      defaultOptions: {
        queries: {
          retry: false
        }
      }
    });

    return (
      <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
    );
  };

  sharedModelTests(ModelShowDescription);

  it(`should allow paths to be directly overridden`, async () => {
    const { asFragment } = render(
      <ModelShowSimple
        model="blog"
        // Default title of null so that empty shows
        extraDefaultValues={{ title: null }}
        fallback={false}
      >
        <ModelShowDescription paths={['title']} />
      </ModelShowSimple>,
      { wrapper }
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
