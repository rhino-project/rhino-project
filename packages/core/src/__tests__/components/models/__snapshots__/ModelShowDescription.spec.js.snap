// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ModelShowDescription > should allow paths to be directly overridden 1`] = `
<DocumentFragment>
  <div
    class="model-show-description model-show-description-blog"
  >
    <div
      class="mb-3"
      id="DisplayLayout-title"
      label="Title"
      path="title"
      placeholder="Title"
      required=""
    >
      <label
        class="required form-label"
        for="title"
        id="displayLabel-title"
        path="title"
        placeholder="Title"
        required=""
      >
        Title
      </label>
      <input
        aria-invalid="false"
        autocomplete="off"
        class="form-control"
        id="title"
        label="Title"
        name="title"
        path="title"
        placeholder="Title"
        readonly=""
        required=""
        type="text"
        value="-"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`ModelShowDescription > should render with global override shorthand 1`] = `
<DocumentFragment>
  <div>
    Bar
  </div>
</DocumentFragment>
`;

exports[`ModelShowDescription > should render with global override shorthand for model 1`] = `
<DocumentFragment>
  <div>
    Bar
  </div>
</DocumentFragment>
`;
